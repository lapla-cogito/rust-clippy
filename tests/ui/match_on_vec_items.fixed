#![warn(clippy::match_on_vec_items)]
#![allow(clippy::redundant_at_rest_pattern, clippy::useless_vec)]

fn match_with_wildcard() {
    let arr = vec![0, 1, 2, 3];
    let range = 1..3;
    let idx = 1;

    // Lint, may panic
    match arr.get(idx) {
        //~^ ERROR: indexing into a vector may panic
        //~| NOTE: `-D clippy::match-on-vec-items` implied by `-D warnings`
        Some(0) => println!("0"),
        Some(1) => println!("1"),
        _ => {},
    }

    // Lint, may panic
    match arr.get(range) {
        //~^ ERROR: indexing into a vector may panic
        Some([0, 1]) => println!("0 1"),
        Some([1, 2]) => println!("1 2"),
        _ => {},
    }
}

fn match_without_wildcard() {
    let arr = vec![0, 1, 2, 3];
    let range = 1..3;
    let idx = 2;

    // Lint, may panic
    match arr.get(idx) {
        //~^ ERROR: indexing into a vector may panic
        Some(0) => println!("0"),
        Some(1) => println!("1"),
        Some(num) => {}
        None => {},
    }

    // Lint, may panic
    match arr.get(range) {
        //~^ ERROR: indexing into a vector may panic
        Some([0, 1]) => println!("0 1"),
        Some([1, 2]) => println!("1 2"),
        Some([ref sub @ ..]) => {}
        None => {},
    }
}

fn match_wildcard_and_action() {
    let arr = vec![0, 1, 2, 3];
    let range = 1..3;
    let idx = 3;

    // Lint, may panic
    match arr.get(idx) {
        //~^ ERROR: indexing into a vector may panic
        Some(0) => println!("0"),
        Some(1) => println!("1"),
        _ => println!("Hello, World!"),
    }

    // Lint, may panic
    match arr.get(range) {
        //~^ ERROR: indexing into a vector may panic
        Some([0, 1]) => println!("0 1"),
        Some([1, 2]) => println!("1 2"),
        _ => println!("Hello, World!"),
    }
}

fn match_vec_ref() {
    let arr = &vec![0, 1, 2, 3];
    let range = 1..3;
    let idx = 3;

    // Lint, may panic
    match arr.get(idx) {
        //~^ ERROR: indexing into a vector may panic
        Some(0) => println!("0"),
        Some(1) => println!("1"),
        _ => {},
    }

    // Lint, may panic
    match arr.get(range) {
        //~^ ERROR: indexing into a vector may panic
        Some([0, 1]) => println!("0 1"),
        Some([1, 2]) => println!("1 2"),
        _ => {},
    }
}

fn match_with_get() {
    let arr = vec![0, 1, 2, 3];
    let range = 1..3;
    let idx = 3;

    // Ok
    match arr.get(idx) {
        Some(0) => println!("0"),
        Some(1) => println!("1"),
        _ => {},
    }

    // Ok
    match arr.get(range) {
        Some(&[0, 1]) => println!("0 1"),
        Some(&[1, 2]) => println!("1 2"),
        _ => {},
    }
}

fn match_with_array() {
    let arr = [0, 1, 2, 3];
    let range = 1..3;
    let idx = 3;

    // Ok
    match arr[idx] {
        0 => println!("0"),
        1 => println!("1"),
        _ => {},
    }

    // Ok
    match arr[range] {
        [0, 1] => println!("0 1"),
        [1, 2] => println!("1 2"),
        _ => {},
    }
}

fn match_with_endless_range() {
    let arr = vec![0, 1, 2, 3];
    let range = ..;

    // Ok
    match arr[range] {
        [0, 1] => println!("0 1"),
        [1, 2] => println!("1 2"),
        [0, 1, 2, 3] => println!("0, 1, 2, 3"),
        _ => {},
    }

    // Ok
    match arr[..] {
        [0, 1] => println!("0 1"),
        [1, 2] => println!("1 2"),
        [0, 1, 2, 3] => println!("0, 1, 2, 3"),
        _ => {},
    }
}

fn or_patterns() {
    let arr = vec![0, 1, 2, 3];
    match arr.get(1) {
        //~^ ERROR: indexing into a vector may panic
        Some(0 | 1) => println!("0 or 1"),
        _ => println!("Hello, World!"),
    }
}

fn main() {
    match_with_wildcard();
    match_without_wildcard();
    match_wildcard_and_action();
    match_vec_ref();
    match_with_get();
    match_with_array();
    match_with_endless_range();
}
