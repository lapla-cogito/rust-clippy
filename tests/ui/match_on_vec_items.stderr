error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:10:11
   |
LL |     match arr[idx] {
   |           ^^^^^^^^
   |
   = note: `-D clippy::match-on-vec-items` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::match_on_vec_items)]`
help: try
   |
LL ~     match arr.get(idx) {
LL |
LL |
LL ~         Some(0) => println!("0"),
LL ~         Some(1) => println!("1"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:19:11
   |
LL |     match arr[range] {
   |           ^^^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(range) {
LL |
LL ~         Some([0, 1]) => println!("0 1"),
LL ~         Some([1, 2]) => println!("1 2"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:33:11
   |
LL |     match arr[idx] {
   |           ^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(idx) {
LL |
LL ~         Some(0) => println!("0"),
LL ~         Some(1) => println!("1"),
LL ~         Some(num) => {}
LL ~         None => {},
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:41:11
   |
LL |     match arr[range] {
   |           ^^^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(range) {
LL |
LL ~         Some([0, 1]) => println!("0 1"),
LL ~         Some([1, 2]) => println!("1 2"),
LL ~         Some([ref sub @ ..]) => {}
LL ~         None => {},
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:55:11
   |
LL |     match arr[idx] {
   |           ^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(idx) {
LL |
LL ~         Some(0) => println!("0"),
LL ~         Some(1) => println!("1"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:63:11
   |
LL |     match arr[range] {
   |           ^^^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(range) {
LL |
LL ~         Some([0, 1]) => println!("0 1"),
LL ~         Some([1, 2]) => println!("1 2"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:77:11
   |
LL |     match arr[idx] {
   |           ^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(idx) {
LL |
LL ~         Some(0) => println!("0"),
LL ~         Some(1) => println!("1"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:85:11
   |
LL |     match arr[range] {
   |           ^^^^^^^^^^
   |
help: try
   |
LL ~     match arr.get(range) {
LL |
LL ~         Some([0, 1]) => println!("0 1"),
LL ~         Some([1, 2]) => println!("1 2"),
   |

error: indexing into a vector may panic
  --> tests/ui/match_on_vec_items.rs:156:11
   |
LL |     match arr[1] {
   |           ^^^^^^
   |
help: try
   |
LL ~     match arr.get(1) {
LL |
LL ~         Some(0 | 1) => println!("0 or 1"),
   |

error: aborting due to 9 previous errors

